# 2. Here’s the list of reviews of Chrome apps - scraped from Playstore.  DataSet Link

**Problem statement - There are times when a user writes Good, Nice App or any other positive text, in the review and gives 1-star rating. Your goal is to identify the reviews where the semantics of review text does not match rating.  
Your goal is to identify such ratings where review text is good, but rating is negative- so that the support team can point this to users.  
Deploy it using - Flask/Streamlit etc and share the live link.  
Important 
In this data app - the user will upload a csv and you would be required to display the reviews where the content doesn’t match ratings.    
This csv will be in the same format as the DataSet Link.   
Bonus Points - If you deploy the app with Authentication.**


In [147]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [4]:
cd ..

D:\


In [5]:
cd Test answers

D:\Test answers


In [6]:
ls

 Volume in drive D has no label.
 Volume Serial Number is 06DA-8863

 Directory of D:\Test answers

05/01/2022  08:55 PM    <DIR>          .
05/01/2022  08:55 PM    <DIR>          ..
04/30/2022  08:35 PM    <DIR>          .ipynb_checkpoints
04/30/2022  06:22 PM         1,984,182 chrome_reviews (1).csv
05/01/2022  08:55 PM            65,341 Untitled.ipynb
               2 File(s)      2,049,523 bytes
               3 Dir(s)  240,199,356,416 bytes free


In [7]:
df=pd.read_csv('chrome_reviews (1).csv')

In [8]:
df.head()

ID                                         Review URL  \
0  3886  https://play.google.com/store/apps/details?id=...   
1  3887  https://play.google.com/store/apps/details?id=...   
2  3888  https://play.google.com/store/apps/details?id=...   
3  3889  https://play.google.com/store/apps/details?id=...   
4  3890  https://play.google.com/store/apps/details?id=...   

                                             Text  Star  Thumbs Up  \
0                      This is very helpfull aap.     5          0   
1                                            Good     3          2   
2  Not able to update. Neither able to uninstall.     1          0   
3                                        Nice app     4          0   
4                               Many unwanted ads     1          0   

          User Name Developer Reply        Version Review Date  \
0  INDIAN Knowledge             NaN  83.0.4103.106  2020-12-19   
1  Ijeoma Happiness             NaN  85.0.4183.127  2020-12-19   
2  Priti D BtCFs-29             NaN  85.0.4183.127  2020-12-19   
3        Ajeet Raja             NaN  77.0.3865.116  2020-12-19   
4           Rams Mp             NaN   87.0.4280.66  2020-12-19   

               App ID  
0  com.android.chrome  
1  com.android.chrome  
2  com.android.chrome  
3  com.android.chrome  
4  com.android.chrome

In [9]:
df.tail()

ID                                         Review URL  \
7199  684987  https://play.google.com/store/apps/details?id=...   
7200  684988  https://play.google.com/store/apps/details?id=...   
7201  684989  https://play.google.com/store/apps/details?id=...   
7202  684990  https://play.google.com/store/apps/details?id=...   
7203  684991  https://play.google.com/store/apps/details?id=...   

                                                   Text  Star  Thumbs Up  \
7199                                          Bagusss..     5          0   
7200                                      Bad version 😔     1          0   
7201  One thing that I have to say I can't spelled t...     5          0   
7202                                          Excellent     5          0   
7203  After update it lag and always slow same goes ...     1          0   

            User Name Developer Reply        Version Review Date  \
7199  boima panjaitan             NaN   88.0.4324.93  2021-03-05   
7200     निशान्त सिंह             NaN  88.0.4324.181  2021-03-05   
7201     Virgie Allen             NaN   89.0.4389.72  2021-03-05   
7202   Kazuo Guevarra             NaN   89.0.4389.72  2021-03-05   
7203       Hanin Amni             NaN   88.0.4324.93  2021-03-05   

                  App ID  
7199  com.android.chrome  
7200  com.android.chrome  
7201  com.android.chrome  
7202  com.android.chrome  
7203  com.android.chrome

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7204 entries, 0 to 7203
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   ID               7204 non-null   int64 
 1   Review URL       7204 non-null   object
 2   Text             7203 non-null   object
 3   Star             7204 non-null   int64 
 4   Thumbs Up        7204 non-null   int64 
 5   User Name        7204 non-null   object
 6   Developer Reply  95 non-null     object
 7   Version          7119 non-null   object
 8   Review Date      7204 non-null   object
 9   App ID           7204 non-null   object
dtypes: int64(3), object(7)
memory usage: 562.9+ KB


In [11]:
df.describe()

ID         Star    Thumbs Up
count    7204.000000  7204.000000  7204.000000
mean   252654.665047     3.592726     1.216685
std    204647.596623     1.728362    41.382835
min      3886.000000     1.000000     0.000000
25%     83254.750000     1.000000     0.000000
50%    181866.500000     5.000000     0.000000
75%    419485.250000     5.000000     0.000000
max    684991.000000     5.000000  2389.000000

In [12]:
#Finding the null values in the dataframe
df.isnull().sum()

ID                    0
Review URL            0
Text                  1
Star                  0
Thumbs Up             0
User Name             0
Developer Reply    7109
Version              85
Review Date           0
App ID                0
dtype: int64

In [13]:
#Dropped the null value in Text column
df.dropna(how='any',subset=['Text'], axis=0, inplace=True)

In [14]:
df.isnull().sum()

ID                    0
Review URL            0
Text                  0
Star                  0
Thumbs Up             0
User Name             0
Developer Reply    7108
Version              85
Review Date           0
App ID                0
dtype: int64

In [32]:
#Finding the length of the test mentioned in review, so that we can analyze the relation of lenth and type of review
df['Text_len']=df['Text'].apply(len)

In [16]:
df.head()

ID                                         Review URL  \
0  3886  https://play.google.com/store/apps/details?id=...   
1  3887  https://play.google.com/store/apps/details?id=...   
2  3888  https://play.google.com/store/apps/details?id=...   
3  3889  https://play.google.com/store/apps/details?id=...   
4  3890  https://play.google.com/store/apps/details?id=...   

                                             Text  Star  Thumbs Up  \
0                      This is very helpfull aap.     5          0   
1                                            Good     3          2   
2  Not able to update. Neither able to uninstall.     1          0   
3                                        Nice app     4          0   
4                               Many unwanted ads     1          0   

          User Name Developer Reply        Version Review Date  \
0  INDIAN Knowledge             NaN  83.0.4103.106  2020-12-19   
1  Ijeoma Happiness             NaN  85.0.4183.127  2020-12-19   
2  Priti D BtCFs-29             NaN  85.0.4183.127  2020-12-19   
3        Ajeet Raja             NaN  77.0.3865.116  2020-12-19   
4           Rams Mp             NaN   87.0.4280.66  2020-12-19   

               App ID  Text_len  
0  com.android.chrome        26  
1  com.android.chrome         4  
2  com.android.chrome        46  
3  com.android.chrome         8  
4  com.android.chrome        17

Analyzed the dataset depending on the length of the review found the below conclusions:

1)Total reviews 7204

2)1 star review 1894

3)2 star review 336

4)3 star review 451

5)4 star review 652

6)5 star review 3871

Found that 3 star rating having average reviews, so that excluded that for analysis.

1) 4 star and 5 star is considered as good review.

2) 1 star and 2 star is considered as bad review.

For 4 and 5 star rating combinely having 4523 reviews. For 1 and 2 ratings combinely having 2230. In that found the logic that if the length of the review is 10 and more than 10 having matching the review with the ratings.

For rating length of word less than 10 having some unmatching reviews.

For 4 & 5 star rating there are around 2594 reviews which are having length less than 10.
Removed the duplicate words from that, it came out to 659.
Collected the words for 659 there are 242 words which are relavent to rating rest 417 are not.
Based on that created logic, kindly refer below for code.

In [17]:
Texts1=['Nice app','Yes yes','Awesome','Good','Suuper','Best','Ok',"it's ok",'Nice','Very good','Excellent','Best app',\
       'nice apps','Perfect','beest','Super','Very nice','Nise','Great','Wow','Love it','Welcome','its ok','Excelente',\
       'Good app','Better','Cool','Super aap','Suppepp','Yeah','Good nice','OK.','Nce app','Nice app.','Tnx','Superbb',\
       'Fantastic','Fine',"It's good",'Superb','Good Apps','Good lahh','It great','Great app','All good','Okay app','Cool cool',\
       'approved','Good one','The best','Cool app','Nice!','Liked','Lovely','awsome','Good nise','Great ty','Nice hot',"Nice 👍",\
       'Okay','Like eat','All great','Supper','4 star','v good','Useful','Its nice''Like it','faster','Nice aap','Reliable',\
       'Sweet','Fast','Great.','So nice',"Good 👍",'Mast',"It's cool",'Awesome.','Helpful','Great!','Its good','hapy','like'\
       'I love it','Lovly','Good app.','perfct','Exalant','Great!!!','I like it','It is ok','Suppar','loved','Good 12','ok app'\
       'Superbbb','Good job','Excellant','Good 9','Cool app.','Super app','Ok good','Oki','Very fine','Thank','Well','I am okk'\
       'Nice work','Vety good','Very best','Gud','Love','Mst','Spr','whaaa','Wadar ful','Nice one','Vari good','Mst. Aao'\
       'Good Best','Awsm app','Yes good','Awesomep','perfect.','Beautiful','Gud app','Its great','Fun','Amazing','Top','Osam'\
       'Okk','Fast ever','Neat',"Superb❤",'Fine one','So Good','Good work','Happy','Suprrr','Enjoy','Yes nice','Nice apss'\
       'Yes Ok','Nice G','I like','Vood','Exlent','best apps','Anazing','super...','Well done','Amezing','Good asp','Good uwu'\
       'Best....',"Good💯",'Good one!','superrrr','Nice apk','B Nice','It good',"Awsm 👌",'Supperb','Great apo','Mast Hai'\
       'Fine app','Good appp','To good','Bas shi h','Good.','Suppr app','Great!!','Woooooow','Amazing.',"Good ðŸ¤",'Good Hghh',\
       'Exelent','Wonderful','just wow',"Best💝",'Nice.','Excelent','Goood','Koool app','Good le','Esy','Good as','This best',\
       '5star',"Nice 👌",'Supar app','Nice web',"Good 😊",'Woh','Aaa great','Nicr','V.gd','Satisfied','Best good','Gooood',\
       'Superhero','Nice up','Niec','good good','only one','nice good','The first','Excilent','Perfecto','Yes cess','Best Aap',\
       'Nice job','Five star','Nice paK','wow app','Well good','Excellen5','Excelenr','Too good','Supare','Usefull','Xlent'\
       'Best h','very well',"🙂 Nice","Nice 🙂",'Must good','Great !','Achha hai',"It's nice",'Sup','Honey spr','Suparb',\
       'Exelleant','This good','Asm','Very niec','Supar','Very fast','Excllent']
filt=df.loc[(~df['Text'].isin(Texts1)) & (df.Text_len < 10) & (df.Star>3)]

In [18]:
filt

ID                                         Review URL       Text  \
25      3911  https://play.google.com/store/apps/details?id=...  Rait crom   
34      3920  https://play.google.com/store/apps/details?id=...    Not bad   
36      3922  https://play.google.com/store/apps/details?id=...  mast xnxx   
58      3944  https://play.google.com/store/apps/details?id=...  Not good!   
61      3947  https://play.google.com/store/apps/details?id=...    So long   
...      ...                                                ...        ...   
7167  684955  https://play.google.com/store/apps/details?id=...         ❤️   
7169  684957  https://play.google.com/store/apps/details?id=...   I'm Goes   
7176  684964  https://play.google.com/store/apps/details?id=...       😊😊😊😊   
7187  684975  https://play.google.com/store/apps/details?id=...     Ponraj   
7199  684987  https://play.google.com/store/apps/details?id=...  Bagusss..   

      Star  Thumbs Up           User Name Developer Reply        Version  \
25       5          0   DJ song full bass             NaN  83.0.4103.106   
34       5          0  Roioehhd Hehksifgh             NaN            NaN   
36       5          0      Badsha boy boy             NaN  36.0.1985.135   
58       4          0       yvonne mojica             NaN  87.0.4280.101   
61       5          0          MANOJ Dora             NaN  85.0.4183.127   
...    ...        ...                 ...             ...            ...   
7167     5          0        ANMOL RAJ AM             NaN  85.0.4183.127   
7169     5          0          Kyei Komla             NaN  81.0.4044.138   
7176     5          0      Shadan Qureshi             NaN   61.0.3163.98   
7187     5          0           Raj Kumar             NaN   54.0.2840.85   
7199     5          0     boima panjaitan             NaN   88.0.4324.93   

     Review Date              App ID  Text_len  
25    2020-12-19  com.android.chrome         9  
34    2020-12-19  com.android.chrome         7  
36    2020-12-19  com.android.chrome         9  
58    2020-12-19  com.android.chrome         9  
61    2020-12-19  com.android.chrome         7  
...          ...                 ...       ...  
7167  2021-03-05  com.android.chrome         2  
7169  2021-03-05  com.android.chrome         8  
7176  2021-03-05  com.android.chrome         4  
7187  2021-03-05  com.android.chrome         6  
7199  2021-03-05  com.android.chrome         9  

[835 rows x 11 columns]

For 1 & 2 star rating there are around 316 reviews which are having length less than 10. Removed the duplicate words from that, it came out to 181 unique reviews. Collected the words for 181 unique reviews.There are 39 words which are relavent to rating, rest 142 are not. Based on that created logic, kindly refer below for code.

In [19]:
Texts2=['Bad','Bad !','Bad ...','Bad app','bad tabs','Bakwas','Bekar','Best app','Better','Cheap','Cool','Cool app','Errors...'\
        'Hate it','Horrible','Low speed','No good','Nonsense','not bad','Not cool','Not fast','Not updat','Not well'\
        'Poor','Slow','Slow work','Very baad','Very bad','Very bed','Very poor','Very slow','Vorest','Waste','Wirst',\
        'Worst','Worst app','Worst!','Wrost app','Yo slow']
filt2=df.loc[(~df['Text'].isin(Texts2)) & (df.Text_len < 10) & (df.Star<3)]

In [20]:
filt2

ID                                         Review URL      Text  \
56      3942  https://play.google.com/store/apps/details?id=...         👎   
101     4113  https://play.google.com/store/apps/details?id=...      Good   
158     4143  https://play.google.com/store/apps/details?id=...      Good   
178     4164  https://play.google.com/store/apps/details?id=...      Help   
226     5186  https://play.google.com/store/apps/details?id=...     Vinay   
...      ...                                                ...       ...   
7060  659674  https://play.google.com/store/apps/details?id=...        👌👌   
7076  659690  https://play.google.com/store/apps/details?id=...    Second   
7111  684899  https://play.google.com/store/apps/details?id=...       Hjj   
7113  684901  https://play.google.com/store/apps/details?id=...       Idk   
7188  684976  https://play.google.com/store/apps/details?id=...  NET slow   

      Star  Thumbs Up            User Name Developer Reply        Version  \
56       1          0         Michael Kuhn             NaN  81.0.4044.111   
101      1          0        Sohail Soomro             NaN  74.0.3729.136   
158      1          0        Md Rubel khan             NaN            NaN   
178      2          0  George & Jodie Wood             NaN   83.0.4103.96   
226      1          0    Ramsurat Ramsurat             NaN            NaN   
...    ...        ...                  ...             ...            ...   
7060     1          0          Sadia Akter             NaN  88.0.4324.181   
7076     2          0        Sh Wijerathna             NaN   55.0.2883.91   
7111     1          0            Ram Ghule             NaN            NaN   
7113     1          0               No you             NaN  88.0.4324.181   
7188     1          0          Afzal Afini             NaN  83.0.4103.101   

     Review Date              App ID  Text_len  
56    2020-12-19  com.android.chrome         1  
101   2020-12-21  com.android.chrome         4  
158   2020-12-21  com.android.chrome         4  
178   2020-12-21  com.android.chrome         4  
226   2020-12-21  com.android.chrome         5  
...          ...                 ...       ...  
7060  2021-03-04  com.android.chrome         2  
7076  2021-03-04  com.android.chrome         6  
7111  2021-03-05  com.android.chrome         3  
7113  2021-03-05  com.android.chrome         3  
7188  2021-03-05  com.android.chrome         8  

[248 rows x 11 columns]

In [21]:
frames= [filt2,filt]

In [22]:
#combined the nonrelavent ratings for 1,2 star and 4,5 star
Non_relavent_reviews = pd.concat(frames)

In [23]:
Non_relavent_reviews

ID                                         Review URL       Text  \
56      3942  https://play.google.com/store/apps/details?id=...          👎   
101     4113  https://play.google.com/store/apps/details?id=...       Good   
158     4143  https://play.google.com/store/apps/details?id=...       Good   
178     4164  https://play.google.com/store/apps/details?id=...       Help   
226     5186  https://play.google.com/store/apps/details?id=...      Vinay   
...      ...                                                ...        ...   
7167  684955  https://play.google.com/store/apps/details?id=...         ❤️   
7169  684957  https://play.google.com/store/apps/details?id=...   I'm Goes   
7176  684964  https://play.google.com/store/apps/details?id=...       😊😊😊😊   
7187  684975  https://play.google.com/store/apps/details?id=...     Ponraj   
7199  684987  https://play.google.com/store/apps/details?id=...  Bagusss..   

      Star  Thumbs Up            User Name Developer Reply        Version  \
56       1          0         Michael Kuhn             NaN  81.0.4044.111   
101      1          0        Sohail Soomro             NaN  74.0.3729.136   
158      1          0        Md Rubel khan             NaN            NaN   
178      2          0  George & Jodie Wood             NaN   83.0.4103.96   
226      1          0    Ramsurat Ramsurat             NaN            NaN   
...    ...        ...                  ...             ...            ...   
7167     5          0         ANMOL RAJ AM             NaN  85.0.4183.127   
7169     5          0           Kyei Komla             NaN  81.0.4044.138   
7176     5          0       Shadan Qureshi             NaN   61.0.3163.98   
7187     5          0            Raj Kumar             NaN   54.0.2840.85   
7199     5          0      boima panjaitan             NaN   88.0.4324.93   

     Review Date              App ID  Text_len  
56    2020-12-19  com.android.chrome         1  
101   2020-12-21  com.android.chrome         4  
158   2020-12-21  com.android.chrome         4  
178   2020-12-21  com.android.chrome         4  
226   2020-12-21  com.android.chrome         5  
...          ...                 ...       ...  
7167  2021-03-05  com.android.chrome         2  
7169  2021-03-05  com.android.chrome         8  
7176  2021-03-05  com.android.chrome         4  
7187  2021-03-05  com.android.chrome         6  
7199  2021-03-05  com.android.chrome         9  

[1083 rows x 11 columns]

non_related_rev

**Found that there are 1083 reviews are not relavent to the rating.**